src/utils/sources-tree/addToTree.js

```diff
@@ -12,14 +14,118 @@ import { getURL, getFilenameFromPath } from "./getURL";
 import type { Node } from "./types";
 import type { SourceRecord } from "../../reducers/types";
 
-function createNodeInTree(part: string, path: string, tree: Node) {
+function createNodeInTree(
+  part: string,
+  path: string,
+  tree: Node,
+  index: number
+) {
   const node = createNode(part, path, []);
+
   // we are modifying the tree
-  tree.contents = [...tree.contents, node];
+  const contents = tree.contents.slice(0);
+  contents.splice(index, 0, node);
+  tree.contents = contents;
+
   return node;
 }
 
 /*
+ * Gets domain from url (without www prefix)
+ */
+function getDomain(url: string): ?string {
+  const { host } = parse(url);
+  if (!host) {
+    return null;
+  }
+  return host.indexOf("www.") === 0 ? host.substr(4) : host;
+}
+
+/*
+ * Checks if node name matches debugger host/domain.
+ */
+function isExactDomainMatch(part: string, debuggeeHost: string): boolean {
+  return part.indexOf("www.") === 0
+    ? part.substr(4) === debuggeeHost
+    : part === debuggeeHost;
+}
+
+type FindNodeInContentsMatcher = any => number;
+/*
+ * Performs a binary search to insert a node into contents. Returns positive
+ * number, index of the found child, or negative number, which can be used
+ * to calculate a position where a new node can be inserted (`-index - 1`).
+ * The matcher is a function that returns result of comparision of a node with
+ * lookup value.
+ */
+function findNodeInContents(tree: Node, matcher: FindNodeInContentsMatcher) {
+  const { contents } = tree;
+  if (contents.length === 0) {
+    return -1;
+  }
+  let l = 0,
+    r = contents.length - 1;
+  while (l < r) {
+    const c = Math.floor((l + r) / 2);
+    if (matcher(contents[c]) < 0) {
+      l = c + 1;
+    } else {
+      r = c;
+    }
+  }
+  const result = matcher(contents[l]);
+  if (result === 0) {
+    return l;
+  }
+  return -(result > 0 ? l : l + 1) - 1;
+}
+
+/*
+ * Creates a matcher for findNodeInContents.
+ * The sorting order of nodes during comparison is:
+ * - "(index)" node
+ * - root node with the debuggee host/domain
+ * - hosts/directories (not files) sorted by name
+ * - files sorted by name
+ */
+function createTreeNodeMatcher(
```

Comment:
this should be moved to its own file, each HOF can be split into its own function and named. that way you wont have to add so many comments

```javascript
function indexMatcher(indexName) {
  return (node) => (node.name === IndexName ? 0 : 1);
}
function domainMatcher(indexName, debuggeeHost) {
 return node => {
   if (node.name === IndexName) {
      return -1;
    }
    return isExactDomainMatch(node.name, (debuggeeHost: any)) ? 0 : 1;
  };
}

createTreeNodeMatcher(...) {
  //...
  if (part === IndexName) {
    return IndexMatcher(indexName)
  } 
  if (debuggeeHost && isExactDomainMatch(part, debuggeeHost)) {
    return domainMatcher(indexName, debuggeeHost)
  }
  // ... etc
```

Reply:
Code was moved to the ./treeOrder.js file.